WcaOnRails/lib/relations.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+module Relations
+  def self.get_chain(wca_id1, wca_id2)
+    find_chain([[wca_id1]], [[wca_id2]])
+  end
+
+  def self.competitions_together(wca_id1, wca_id2)
+    sql = <<-SQL
+      SELECT competition.id, competition.name
+      FROM people_pairs_with_competition
+      JOIN Competitions competition ON competition.id = competition_id
+      WHERE wca_id1 = '#{wca_id1}' AND wca_id2 = '#{wca_id2}'
+    SQL
+    ActiveRecord::Base.connection.execute(sql).to_a
+  end
+
+  # Internal
+
+  def self.find_chain(left_chains, right_chains)
+    extended_chains_by_one_degree!(left_chains)
+    final_chain = random_linked_chain(left_chains, right_chains)
+    return final_chain if final_chain
+    extended_chains_by_one_degree!(right_chains)
+    final_chain = random_linked_chain(left_chains, right_chains)
+    return final_chain if final_chain
+    find_chain(left_chains, right_chains)
+  end
+
+  def self.extended_chains_by_one_degree!(chains)
+    chains.map! do |chain|
+      linkings[chain.last].map { |wca_id| [*chain, wca_id] }
+    end.flatten!(1)
+  end
+
+  def self.random_linked_chain(left_chains, right_chains)
+    for *left_chain, left_last in left_chains.shuffle
+      for *right_chain, right_last in right_chains.shuffle
+        return [*left_chain, left_last, *right_chain.reverse] if left_last == right_last
+      end
+    end
+    nil
+  end
+
+  def self.linkings
+    @@linkings ||= ActiveRecord::Base.connection.execute("SELECT wca_id, wca_ids FROM linkings")
+                                     .to_a.map! { |wca_id, wca_ids| [wca_id, wca_ids.split(',')] }
+                                     .to_h.freeze
+  end
+
+  def self.compute_auxiliary_data
+    sql = File.read(Rails.root.join('lib', 'relations_compute_auxiliary_data.sql'))
+    sql.split(';').each do |statement|
```

Comment:
Oh dear, this seems potentially problematic. If someone ever adds a quoted string with a semicolon in it, it's going to be sad times =( I had to deal with this when writing the database_dumper_spec. Maybe the solution in that test could help you? See [this hack](https://github.com/jfly/worldcubeassociation.org/blob/6b8a42146874f699f1970ff9397cff50cbae23eb/WcaOnRails/spec/lib/database_dumper_spec.rb#L11), and then [this hack](https://github.com/jfly/worldcubeassociation.org/blob/6b8a42146874f699f1970ff9397cff50cbae23eb/WcaOnRails/spec/lib/database_dumper_spec.rb#L52-L56). Maybe not worth implementing...